### PR TITLE
Adding a new topic of how to add a new dynamic rows system config

### PR DESCRIPTION
**Pull request by eduard13**
## Purpose of this pull request

This pull request (PR) adds a new tutorial where it is described how to add a dynamic rows system config into Magento backend. Also provides a preview of the new added system config.

## Affected DevDocs pages

- https://devdocs.magento.com/guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.html
- https://devdocs.magento.com/guides/v2.2/ext-best-practices/tutorials/dynamic-row-system-config.html
- https://devdocs.magento.com/guides/v2.3/ext-best-practices/tutorials/dynamic-row-system-config.html

whatsnew
Added a [new tutorial](https://devdocs.magento.com/guides/v2.3/ext-best-practices/tutorials/dynamic-row-system-config.html) that shows you how to add a dynamic rows system configuration

### CODE REVIEW

**Comment by devops-devdocs** [comment]
An admin must run tests on this PR before it can be merged.

**Comment by jeff-matthews** [comment]
I'm excited to see a new topic contribution @eduard13! Thank you! After @osrecio does a technical review, we'll do an editorial review.

File: guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+group: extension-best-practices
+subgroup: 02_Extension-Coding
+title: Creating a dynamic row system config
+menu_title: Creating a dynamic row system config
+menu_order: 1010
+functional_areas:
+  - Standards
+---
+
+### Overview
```
Comment (shrielenee) [inline guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md]: Remove `### Overview` heading. We don't need these.

File: guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+This tutorial gives instructions for adding a new dynamic rows system config in the {% glossarytooltip 18b930cf-09cc-47c9-a5e5-905f86c43f81 %}Magento admin{% endglossarytooltip %}, by extending the [Magento/Config/Block/System/Config/Form/Field/FieldArray/AbstractFieldArray][0]{:target="_blank"} class.
```
Comment (shrielenee) [inline guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md]:
```suggestion
This tutorial shows you how to add a new dynamic rows system configuration in the {% glossarytooltip 18b930cf-09cc-47c9-a5e5-905f86c43f81 %}Magento admin{% endglossarytooltip %} by extending the [`Magento/Config/Block/System/Config/Form/Field/FieldArray/AbstractFieldArray`][0]{:target="_blank"} class. 
```

File: guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+---
+
+### Overview
+
+This tutorial gives instructions for adding a new dynamic rows system config in the {% glossarytooltip 18b930cf-09cc-47c9-a5e5-905f86c43f81 %}Magento admin{% endglossarytooltip %}, by extending the [Magento/Config/Block/System/Config/Form/Field/FieldArray/AbstractFieldArray][0]{:target="_blank"} class. 
+
+## Step 1: Add your new system field 
+{:#step-1}
+
+**etc/adminhtml/system.xml**
```
Comment (shrielenee) [inline guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md]:
```suggestion
> `etc/adminhtml/system.xml`
```

File: guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+menu_order: 1010
+functional_areas:
+  - Standards
+---
+
+### Overview
+
+This tutorial gives instructions for adding a new dynamic rows system config in the {% glossarytooltip 18b930cf-09cc-47c9-a5e5-905f86c43f81 %}Magento admin{% endglossarytooltip %}, by extending the [Magento/Config/Block/System/Config/Form/Field/FieldArray/AbstractFieldArray][0]{:target="_blank"} class. 
+
+## Step 1: Add your new system field
```
Comment (shrielenee) [inline guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md]:
```suggestion
## Step 1: Add a new system field 
```

File: guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+                    <frontend_model>[vendor]\[module]\Block\Adminhtml\Form\Field\Ranges</frontend_model>
+                    <backend_model>Magento\Config\Model\Config\Backend\Serialized\ArraySerialized</backend_model>
+                </field>
+            </group>
+        </section>
+    </system>
+</config>
+```
+
+This code adds a new system config on the following location:
```
Comment (shrielenee) [inline guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md]:
```suggestion
This code adds a new system config in the following location:
```

Comment (shrielenee) [reply guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md]: Also remove the spaces between the words and / in the path below.

File: guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+``Stores / Configuration / General [General] / Quantity Ranges``
+
+## Step 2: Create the Frontend model class
+{:#step-2}
+
+**Block/Adminhtml/Form/Field/Ranges.php**
```
Comment (shrielenee) [inline guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md]:
```suggestion
> `Block/Adminhtml/Form/Field/Ranges.php`
```

File: guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        </section>
+    </system>
+</config>
+```
+
+This code adds a new system config on the following location:
+
+``Stores / Configuration / General [General] / Quantity Ranges``
+
+## Step 2: Create the Frontend model class
```
Comment (shrielenee) [inline guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md]:
```suggestion
## Step 2: Create the front-end model class
```

File: guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+        $this->addColumn('from_qty', ['label' => __('From'), 'class' => 'required-entry']);
+        $this->addColumn('to_qty', ['label' => __('To'), 'class' => 'required-entry']);
+        $this->addColumn('price', ['label' => __('Price'), 'class' => 'required-entry']);
+        $this->_addAfter = false;
+        $this->_addButtonLabel = __('Add');
+    }
+}
+```
+
+This block prepares the columns that you want to have in your new config.
```
Comment (shrielenee) [inline guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md]:
```suggestion
This block prepares the desired columns for inclusion in the new config. 
```

File: guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+        $this->_addButtonLabel = __('Add');
+    }
+}
+```
+
+This block prepares the columns that you want to have in your new config. 
+
+### Result
+
+The result is a new system dynamic row field into Admin panel.
```
Comment (shrielenee) [inline guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md]:
```suggestion
The result is a new dynamic system row field in the Admin panel.
```

File: guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+        $this->addColumn('price', ['label' => __('Price'), 'class' => 'required-entry']);
+        $this->_addAfter = false;
+        $this->_addButtonLabel = __('Add');
+    }
+}
+```
+
+This block prepares the columns that you want to have in your new config. 
+
+### Result
```
Comment (shrielenee) [inline guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md]: Should `### Result` be a level 2 heading, because it's the result of both steps?

File: guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+functional_areas:
+  - Standards
+---
+
+### Overview
+
+This tutorial gives instructions for adding a new dynamic rows system config in the {% glossarytooltip 18b930cf-09cc-47c9-a5e5-905f86c43f81 %}Magento admin{% endglossarytooltip %}, by extending the [Magento/Config/Block/System/Config/Form/Field/FieldArray/AbstractFieldArray][0]{:target="_blank"} class. 
+
+## Step 1: Add your new system field 
+{:#step-1}
```
Comment (shrielenee) [inline guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md]: Why are these here for the steps?

`{:#step-1}`, `{:#step-2}`?

Comment (eduard13) [reply guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md]: I was checking another example (in order to show them as steps in the right sidebar), but most probably they are old versioned, so removed them.
Thank you.

File: guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md (hunk 10 above)
Comment (eduard13) [reply guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md]: Agree.
Thank you.

**Comment by eduard13** [comment]
Hi @shrielenee, thank you for your review, could you please check the latest changes?
Thank you a lot.

File: guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+                </field>
+            </group>
+        </section>
+    </system>
+</config>
+```
+
+This code adds a new system config in the following location:
+
+``Stores/Configuration/General [General]/Quantity Ranges``
```
Comment (shrielenee) [inline guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md]: This path doesn't read right, and I think there are too many ` (remove one from each side).

File: guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+
+This block prepares the desired columns for inclusion in the new config.
+
+## Result
+
+The result is a new dynamic system row field in the Admin panel.
+
+![Dynamic Rows System Config]({{ site.baseurl }}/common/images/ext-best-practices/dynamic-rows-config-result.png)
+
+[0]:{{ site.mage2bloburl }}/2.1/app/code/Magento/Config/Block/System/Config/Form/Field/FieldArray/AbstractFieldArray.php
```
Comment (shrielenee) [inline guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md]: I believe there needs to be a space after the `[0]:` for it to render properly.

File: guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md (hunk 12 above)
Comment (eduard13) [reply guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md]: The ` have been removed.

File: guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md (hunk 13 above)
Comment (eduard13) [reply guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md]: I was able to find several examples without space, but anyway I've removed it.

File: guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md (hunk 5 above)
Comment (shrielenee) [reply guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md]: The path isn't showing correctly. When I click View File, I see this:

Screen Shot 2019-05-21 at 8 05 05 AM

File: guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md (hunk 12 above)
Comment (shrielenee) [reply guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md]: Screen Shot 2019-05-21 at 8 05 05 AM

**Comment by shrielenee** [comment]
@eduard13 I have one more question about the way a path is rendering, as seen above in my comments. Please review and edit as necessary, then we should be ready to merge. 🙇‍♀

File: guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md (hunk 12 above)
Comment (eduard13) [reply guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md]: I don't really get why the current rendering is not good, like
`Stores/Configuration/General [General]/Quantity Ranges` ?
Maybe we should use a different separator, like `->`. What do you think?

Comment (shrielenee) [reply guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md]: Why is `General` in brackets? It doesn't make sense. It should be `Stores/Configuration/General/QuantityRanges` (also with no spaces, as it's unlikely the actual path has spaces---am I wrong?)

Is this the actual directory of a file, or a clickable path to get somewhere in the UI?

Directory: https://wiki.corp.magento.com/display/MDOC/OMS+Retros
Clickable path: `**Stores** > **Configuration** > **General** > **QuantityRanges**

Comment (eduard13) [reply guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md]: Regarding the `General` in brackets, I wanted to specify the tab, but agree that it is not the best option. After checking some other examples, looks like a better solution is to show it as:
`**Stores** > **Configuration** > **General** > **Quantity Ranges**`
I was checking this one as example:
https://devdocs.magento.com/guides/v2.3/frontend-dev-guide/templates/template-email.html#customize-content

Comment (shrielenee) [reply guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md]: So, I don't have the UI open, but things that you actually click on should be bolded, and sections/areas that options lie within (but that don't require clicking) should not be bolded. The words should also be capitalized and spelled as exactly as in the UI.

Can you confirm that this applies to your changes to the clickable path?

Comment (eduard13) [reply guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md]: Yes, it is exactly as it is described. The clickable menu items are bolded and the sections (tabs) are not. Also the menu items are spelled and capitalized as they actually are.
Here the green highlighted are clickable items, and the blue ones are not.
Xnip2019-05-21_21-58-21

Comment (shrielenee) [reply guides/v2.1/ext-best-practices/tutorials/dynamic-row-system-config.md]: Great! Thank you for this 💯

**Comment by shrielenee** [comment]
running tests

**Comment by shrielenee** [comment]
running tests

**Comment by shrielenee** [comment]
Great job @eduard13!

**Comment by eduard13** [comment]
Thank you @shrielenee for your help.
